File: Practicas/Qytetet/QytetetJava/src/controladorqytetet/ControladorQytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package controladorqytetet;

import java.util.ArrayList;
import modeloqytetet.MetodoSalirCarcel;
import modeloqytetet.EstadoJuego;
import modeloqytetet.Qytetet;
import modeloqytetet.Jugador;
import modeloqytetet.Casilla;
import java.util.Collections;
/**
 *
 * @author enrique
 */
public class ControladorQytetet {
    
    private ArrayList<String> nombreJugadores = new ArrayList<>();
    private Qytetet modelo = new Qytetet();
    public static final ControladorQytetet instance = new ControladorQytetet();
    
    public ControladorQytetet(){
        
    }
      
    public static ControladorQytetet getInstance()
    {
        return instance;
    }
    
    public void setNombreJugadores(ArrayList<String> nombreJugadores)
    {
        for (String nombre: nombreJugadores)
        {
            this.nombreJugadores.add(nombre);
        }
    }
    
    public ArrayList<Integer> obtenerOperacionesJuegoValidas()
    {
        ArrayList<Integer> operacionesValidas = new ArrayList<>();
        if(modelo.getJugadores().isEmpty())
        {
            operacionesValidas.add(OpcionMenu.INICIARJUEGO.ordinal());
        }
        else
        {
            EstadoJuego estadoJuego = modelo.getEstadoJuego();
            operacionesValidas.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
            operacionesValidas.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
            operacionesValidas.add(OpcionMenu.MOSTRARTABLERO.ordinal());
            operacionesValidas.add(OpcionMenu.TERMINARJUEGO.ordinal());
            switch(estadoJuego)
            {
                case JA_CONSORPRESA:
                {
                    operacionesValidas.add(OpcionMenu.APLICARSORPRESA.ordinal());
                    break;
                }
                case ALGUNJUGADORENBANCARROTA:
                {
                    operacionesValidas.add(OpcionMenu.OBTENERRANKING.ordinal());
                    break;
                }
                case JA_PUEDECOMPRAROGESTIONAR:
                {
                    operacionesValidas.add(OpcionMenu.PASARTURNO.ordinal());
                    operacionesValidas.add(OpcionMenu.VENDERPROPIEDAD.ordinal());
                    operacionesValidas.add(OpcionMenu.COMPRARTITULOPROPIEDAD.ordinal());
                    operacionesValidas.add(OpcionMenu.HIPOTECARPROPIEDAD.ordinal());
                    operacionesValidas.add(OpcionMenu.CANCELARHIPOTECA.ordinal());
                    operacionesValidas.add(OpcionMenu.EDIFICARCASA.ordinal());
                    operacionesValidas.add(OpcionMenu.EDIFICARHOTEL.ordinal());
                    break;
                }
                case JA_PUEDEGESTIONAR:
                {
                    operacionesValidas.add(OpcionMenu.PASARTURNO.ordinal());
                    operacionesValidas.add(OpcionMenu.VENDERPROPIEDAD.ordinal());
                    operacionesValidas.add(OpcionMenu.HIPOTECARPROPIEDAD.ordinal());
                    operacionesValidas.add(OpcionMenu.CANCELARHIPOTECA.ordinal());
                    operacionesValidas.add(OpcionMenu.EDIFICARCASA.ordinal());
                    operacionesValidas.add(OpcionMenu.EDIFICARHOTEL.ordinal());
                    break;
                }
                case JA_PREPARADO:
                {
                    operacionesValidas.add(OpcionMenu.JUGAR.ordinal());
                    break;
                }
                case JA_ENCARCELADO:
                {
                    operacionesValidas.add(OpcionMenu.PASARTURNO.ordinal());
                    break;
                }
                case JA_ENCARCELADOCONOPCIONDELIBERTAD:
                {
                    operacionesValidas.add(OpcionMenu.INTENTARSALIRCARCELPAGANDOLIBERTAD.ordinal());
                    operacionesValidas.add(OpcionMenu.INTENTARSALIRCARCELTIRANDODADO.ordinal());
                    break;
                }
            }
            Collections.sort(operacionesValidas);
        }
        return operacionesValidas;
    }
    
    public boolean necesitaElegirCasilla(int opcionMenu)
    {
        if(opcionMenu == OpcionMenu.HIPOTECARPROPIEDAD.ordinal() || 
           opcionMenu == OpcionMenu.CANCELARHIPOTECA.ordinal()   ||
           opcionMenu == OpcionMenu.EDIFICARCASA.ordinal()       ||
           opcionMenu == OpcionMenu.EDIFICARHOTEL.ordinal()      ||
           opcionMenu == OpcionMenu.VENDERPROPIEDAD.ordinal())
            return true;
        return false;
    }
    
    public ArrayList<Integer> obtenerCasillasValidas(int opcionMenu)
    {
        ArrayList<Integer> propiedades = new ArrayList<>();
        if(opcionMenu == OpcionMenu.HIPOTECARPROPIEDAD.ordinal())
            propiedades = modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(false);
        else if(opcionMenu == OpcionMenu.CANCELARHIPOTECA.ordinal())
            propiedades = modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(true);
        else if(opcionMenu == OpcionMenu.EDIFICARCASA.ordinal()  ||
                opcionMenu == OpcionMenu.EDIFICARHOTEL.ordinal() ||
                opcionMenu == OpcionMenu.VENDERPROPIEDAD.ordinal())
            propiedades = modelo.obtenerPropiedadesJugador();
        return propiedades;
    }
    
    public String realizarOperacion(int opcionElegida, int casillaElegida)
    {
        OpcionMenu opcionMenu = OpcionMenu.values()[opcionElegida];
        String mensaje = "";
        switch(opcionMenu)
        {
            case INICIARJUEGO:
            {
                mensaje =  "Se inicializa el juego";
                modelo.inicializarJuego(nombreJugadores);
                mensaje += "\nEmpieza el jugador " + modelo.getJugadorActual().toString();
                break;
            }
            case JUGAR:
            {
                Casilla casilla_inicial = modelo.getJugadorActual().getCasillaActual();
                mensaje = "Se tira el dado y sale un " + modelo.jugar();
                Casilla casilla_final = modelo.getJugadorActual().getCasillaActual();
                mensaje += "\nEl jugador se mueve de la casilla " + casilla_inicial.toString() +
                           " a la casilla " + casilla_final.toString();
                if(casilla_final.getNumeroCasilla() < casilla_inicial.getNumeroCasilla())
                    mensaje += "\nAdemás, el jugador pasa por la casilla de salida, " +
                    "por lo que cobra " + Qytetet.SALDO_SALIDA;
                break;
            }
            case APLICARSORPRESA:
            {
                mensaje = "Se saca la carta sorpresa:\n" + modelo.getCartaActual().toString();
                modelo.aplicarSorpresa();
                break;
            }
            case INTENTARSALIRCARCELPAGANDOLIBERTAD:
            {
                if(modelo.intentarSalirCarcel(MetodoSalirCarcel.PAGANDOLIBERTAD))
                    mensaje = "El jugador paga " + Qytetet.PRECIO_LIBERTAD + " por lo que sale de la cárcel";
                else
                    mensaje = "El jugador no tiene suficiente dinero para pagar la libertad";
                break;
            }
            case INTENTARSALIRCARCELTIRANDODADO:
            {
                if(modelo.intentarSalirCarcel(MetodoSalirCarcel.TIRANDODADO))
                    mensaje = "El jugador tira un dado y saca más de un 5, por lo que sale de la cárcel";
                else
                    mensaje =  "El jugador tira un dado y saca menos de un 5, por lo que no sale de la cárcel";
                break;
            }
            case COMPRARTITULOPROPIEDAD:
            {
                if(modelo.comprarTituloPropiedad())
                    mensaje = "El jugador compra la casilla";
                else
                    mensaje = "El jugador no tiene saldo suficiente para comprar la casilla";
                break;
            }
            case HIPOTECARPROPIEDAD:
            {
                if(modelo.hipotecarPropiedad(casillaElegida))
                    mensaje = "El jugador hipoteca la casilla elegida";
                else
                    mensaje = "No se puede hipotecar la casilla elegida";
                break;
            }
            case CANCELARHIPOTECA:
            {
                if(modelo.cancelarHipoteca(casillaElegida))
                    mensaje = "El jugador cancela la hipoteca de la casilla elegida";
                else
                    mensaje = "No se puede cancelar la hipoteca de la casilla elegida";
                break;
            }
            case EDIFICARCASA:
            {
                if(modelo.edificarCasa(casillaElegida))
                    mensaje = "El jugador edifica una casa en la casilla elegida";
                else
                    mensaje = "No se puede edificar una casa en la casilla elegida";
                break;
            }
            case EDIFICARHOTEL:
            {
                if(modelo.edificarHotel(casillaElegida))
                    mensaje = "El jugador edifica un hotel en la casilla elegida";
                else
                    mensaje = "No se puede edificar un hotel en la casilla elegida";
                break;
            }
            case VENDERPROPIEDAD:
            {
                if(modelo.venderPropiedad(casillaElegida))
                    mensaje = "El jugador vende la casilla elegida";
                else
                    mensaje = "No se puede vender la casilla elegida";
                break;
            }
            case PASARTURNO:
            {
                modelo.siguienteJugador();
                mensaje = "Se pasa el turno al jugador " + modelo.getJugadorActual().toString();
                break;
            }            
            case OBTENERRANKING:
            {
                mensaje = modelo.obtenerRanking();
                break;
            }  
            case MOSTRARTABLERO:
            {
                mensaje = modelo.getTablero().toString();
                break;
            }
            case MOSTRARJUGADORACTUAL:
            {
                mensaje = modelo.getJugadorActual().toString();
                break;
            }
            case MOSTRARJUGADORES:
            {
                mensaje = "Listado de jugadores:";
                for(Jugador j:modelo.getJugadores())
                    mensaje+= "\n" + j.toString();
                break;
            }
            case TERMINARJUEGO:
            {
                mensaje = "Partida abortada por el jugador";
                break;
            }
        }
        return mensaje;
    }
}

File: Practicas/Qytetet/QytetetRuby/lib/controlador_qytetet.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "opcion_menu"
require_relative "qytetet"
require_relative "estado_juego"
require_relative "metodo_salir_carcel"
require_relative "jugador"
require_relative "casilla"

require "singleton"

module ControladorQytetet
  class ControladorQytetet
    attr_writer :nombre_jugadores
    include Singleton

    def initialize
      @modelo = ModeloQytetet::Qytetet.instance
      @nombre_jugadores = Array.new
    end

    def obtener_operaciones_juego_validas
      operaciones_validas = Array.new
      if(@modelo.jugadores.empty?)
        operaciones_validas << OpcionMenu.index(:INICIARJUEGO)
      else
        estado_juego = @modelo.estado_juego
        operaciones_validas << OpcionMenu.index(:MOSTRARJUGADORACTUAL)
        operaciones_validas << OpcionMenu.index(:MOSTRARJUGADORES)
        operaciones_validas << OpcionMenu.index(:MOSTRARTABLERO)
        operaciones_validas << OpcionMenu.index(:TERMINARJUEGO)
        case estado_juego
          when ModeloQytetet::EstadoJuego::JA_CONSORPRESA
            operaciones_validas << OpcionMenu.index(:APLICARSORPRESA)
          when ModeloQytetet::EstadoJuego::ALGUNJUGADORENBANCARROTA
            operaciones_validas << OpcionMenu.index(:OBTENERRANKING)
          when ModeloQytetet::EstadoJuego::JA_PUEDECOMPRAROGESTIONAR
            operaciones_validas << OpcionMenu.index(:PASARTURNO)
            operaciones_validas << OpcionMenu.index(:VENDERPROPIEDAD)
            operaciones_validas << OpcionMenu.index(:COMPRARTITULOPROPIEDAD)
            operaciones_validas << OpcionMenu.index(:HIPOTECARPROPIEDAD)
            operaciones_validas << OpcionMenu.index(:CANCELARHIPOTECA)
            operaciones_validas << OpcionMenu.index(:EDIFICARCASA)
            operaciones_validas << OpcionMenu.index(:EDIFICARHOTEL)
          when ModeloQytetet::EstadoJuego::JA_PUEDEGESTIONAR
            operaciones_validas << OpcionMenu.index(:PASARTURNO)
            operaciones_validas << OpcionMenu.index(:VENDERPROPIEDAD)
            operaciones_validas << OpcionMenu.index(:HIPOTECARPROPIEDAD)
            operaciones_validas << OpcionMenu.index(:CANCELARHIPOTECA)
            operaciones_validas << OpcionMenu.index(:EDIFICARCASA)
            operaciones_validas << OpcionMenu.index(:EDIFICARHOTEL)
          when ModeloQytetet::EstadoJuego::JA_PREPARADO
            operaciones_validas << OpcionMenu.index(:JUGAR)
          when ModeloQytetet::EstadoJuego::JA_ENCARCELADO
            operaciones_validas << OpcionMenu.index(:PASARTURNO)
          when ModeloQytetet::EstadoJuego::JA_ENCARCELADOCONOPCIONDELIBERTAD
            operaciones_validas << OpcionMenu.index(:INTENTARSALIRCARCELPAGANDOLIBERTAD)
            operaciones_validas << OpcionMenu.index(:INTENTARSALIRCARCELTIRANDODADO)
        end
      end
      operaciones_validas.sort!
      operaciones_validas
    end

    def necesita_elegir_casilla(opcion_menu)
      necesita = false
      if  (opcion_menu == OpcionMenu.index(:HIPOTECARPROPIEDAD) || 
           opcion_menu == OpcionMenu.index(:CANCELARHIPOTECA)   ||
           opcion_menu == OpcionMenu.index(:EDIFICARCASA)       ||
           opcion_menu == OpcionMenu.index(:EDIFICARHOTEL)      ||
           opcion_menu == OpcionMenu.index(:VENDERPROPIEDAD))
        necesita = true
      end
        necesita
    end

    def obtener_casillas_validas(opcion_menu)
      propiedades = Array.new
      if(opcion_menu == OpcionMenu.index(:HIPOTECARPROPIEDAD))
        propiedades = @modelo.obtener_propiedades_jugador_segun_estado_hipoteca(false)
      elsif(opcion_menu == OpcionMenu.index(:CANCELARHIPOTECA))
        propiedades = @modelo.obtener_propiedades_jugador_segun_estado_hipoteca(true)
      elsif(opcion_menu == OpcionMenu.index(:EDIFICARCASA)   ||
            opcion_menu == OpcionMenu.index(:EDIFICARHOTEL) ||
            opcion_menu == OpcionMenu.index(:VENDERPROPIEDAD))
        propiedades = @modelo.obtener_propiedades_jugador
      end
      propiedades
    end

    def realizar_operacion(opcion_elegida, casilla_elegida)
      opcion_menu = OpcionMenu.at(opcion_elegida)

      case opcion_menu
        when :INICIARJUEGO
          mensaje =  "Se inicializa el juego"
          @modelo.inicializar_juego(@nombre_jugadores)
          mensaje += "\nEmpieza el jugador " + @modelo.jugador_actual.to_s
        when :JUGAR
          casilla_inicial = @modelo.jugador_actual.casilla_actual
          mensaje = "Se tira el dado y sale un " + @modelo.jugar.to_s
          casilla_final = @modelo.jugador_actual.casilla_actual
          mensaje += "\nEl jugador se mueve de la casilla " + casilla_inicial.to_s +
                     " a la casilla " + casilla_final.to_s
          if(casilla_final.numero_casilla < casilla_inicial.numero_casilla)
            mensaje += "\nAdemas, el jugador pasa por la casilla de salida, por lo que cobra "
          end
        when :APLICARSORPRESA
          mensaje = "Se saca la carta sorpresa:\n" + @modelo.carta_actual.to_s
          @modelo.aplicar_sorpresa
        when :INTENTARSALIRCARCELPAGANDOLIBERTAD
          if(@modelo.intentar_salir_carcel(ModeloQytetet::MetodoSalirCarcel::PAGANDOLIBERTAD))
            mensaje = "El jugador paga, por lo que sale de la carcel"
          else
            mensaje = "El jugador no tiene suficiente dinero para pagar la libertad"
          end
        when :INTENTARSALIRCARCELTIRANDODADO
          if(@modelo.intentar_salir_carcel(ModeloQytetet::MetodoSalirCarcel::TIRANDODADO))
            mensaje = "El jugador tira un dado y saca mas de un 5, por lo que sale de la carcel"
          else
            mensaje =  "El jugador tira un dado y saca menos de un 5, por lo que no sale de la carcel"
          end
        when :COMPRARTITULOPROPIEDAD
          if(@modelo.comprar_titulo_propiedad)
            mensaje = "El jugador compra la casilla"
          else
            mensaje = "El jugador no tiene saldo suficiente para comprar la casilla"
          end
        when :HIPOTECARPROPIEDAD
          if(@modelo.hipotecar_propiedad(casilla_elegida))
            mensaje = "El jugador hipoteca la casilla elegida"
          else
            mensaje = "No se puede hipotecar la casilla elegida"
          end
        when :CANCELARHIPOTECA
          if(@modelo.cancelar_hipoteca(casilla_elegida))
            mensaje = "El jugador cancela la hipoteca de la casilla elegida"
          else
            mensaje = "No se puede cancelar la hipoteca de la casilla elegida"
          end
        when :EDIFICARCASA
          if(@modelo.edificar_casa(casilla_elegida))
            mensaje = "El jugador edifica una casa en la casilla elegida"
          else
            mensaje = "No se puede edificar una casa en la casilla elegida"
          end
        when :EDIFICARHOTEL
          if(@modelo.edificar_hotel(casilla_elegida))
            mensaje = "El jugador edifica un hotel en la casilla elegida"
          else
            mensaje = "No se puede edificar un hotel en la casilla elegida"
          end
        when :VENDERPROPIEDAD
          if(@modelo.vender_propiedad(casilla_elegida))
            mensaje = "El jugador vende la casilla elegida"
          else
            mensaje = "No se puede vender la casilla elegida"
          end
        when :PASARTURNO
          @modelo.siguiente_jugador
          mensaje = "Se pasa el turno al jugador " + @modelo.jugador_actual.to_s
        when :OBTENERRANKING
          mensaje = @modelo.obtener_ranking
        when :MOSTRARTABLERO
          mensaje = @modelo.tablero.to_s
        when :MOSTRARJUGADORACTUAL
          mensaje = @modelo.jugador_actual.to_s
        when :MOSTRARJUGADORES
          mensaje = "Listado de jugadores:"
          @modelo.jugadores.each do |j|
            mensaje += "\n" + j.to_s
          end
        when :TERMINARJUEGO
          mensaje = "Partida abortada por el jugador"
      end
      mensaje
    end
  end
end
    